File: core/src/com/alta_v2/mediator/ProcessMediatorImpl.java
package com.alta_v2.mediator;

import com.alta_v2.game.ScreenManager;
import com.alta_v2.game.utils.ChangeScreenResult;
import com.alta_v2.mediator.screen.context.ContextFactory;
import com.alta_v2.mediator.screen.context.ScreenContext;
import com.alta_v2.model.MenuDefinitionModel;
import com.alta_v2.model.TiledMapDefinitionModel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import javax.inject.Inject;

/**
 * Provides the mediator that responsible for orchestration of game.
 */
@RequiredArgsConstructor(onConstructor = @__(@Inject))
public class ProcessMediatorImpl implements ProcessMediator {

    private final ContextFactory contextFactory;
    private final ScreenManager screenManager;

    @Getter
    private ScreenContext currentContext;

    /**
     * {@inheritDoc}
     */
    @Override
    public ChangeScreenResult loadMenuScreen(MenuDefinitionModel definition) {
        this.currentContext = this.contextFactory.createMenuContext(definition);
        return this.screenManager.changeScreen(this.currentContext);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ChangeScreenResult loadTiledMapScreen(TiledMapDefinitionModel definition) {
        this.currentContext = this.contextFactory.createTiledMapContext(definition);
        return this.screenManager.changeScreen(this.currentContext);
    }
}

File: core/src/com/alta_v2/mediator/ProcessMediatorImpl.kt
package com.alta_v2.mediator

import com.alta_v2.game.ScreenManager
import com.alta_v2.game.utils.ChangeScreenResult
import com.alta_v2.mediator.screen.context.ContextFactory
import com.alta_v2.mediator.screen.context.ScreenContext
import com.alta_v2.model.MenuDefinitionModel
import com.alta_v2.model.TiledMapDefinitionModel
import com.google.inject.Inject

/**
 * Provides the mediator that responsible for orchestration of game.
 */
class ProcessMediatorImpl @Inject constructor(private val contextFactory: ContextFactory,
                                              private val screenManager: ScreenManager) : ProcessMediator {

    private var currentContext: ScreenContext? = null

    override fun getCurrentContext(): ScreenContext? = currentContext

    /**
     * {@inheritDoc}
     */
    override fun loadMenuScreen(definition: MenuDefinitionModel?): ChangeScreenResult {
        currentContext = contextFactory.createMenuContext(definition!!)
        return screenManager.changeScreen(currentContext)
    }

    /**
     * {@inheritDoc}
     */
    override fun loadTiledMapScreen(definition: TiledMapDefinitionModel?): ChangeScreenResult {
        currentContext = contextFactory.createTiledMapContext(definition!!)
        return screenManager.changeScreen(currentContext)
    }
}